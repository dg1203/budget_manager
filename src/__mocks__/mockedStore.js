import { createStore } from "redux";
import rootReducer from "../reducers/rootReducer";

const initialState = {
  jars: [
    {
      id: "dasas",
      currency: "",
      name: "Test",
      description: "Test",
      resources: []
    }
  ],
  logs: [],
  defaultJar: "fds",
  message: null
};

const store = createStore(rootReducer, initialState);

export default store;
