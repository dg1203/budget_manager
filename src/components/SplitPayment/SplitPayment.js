import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import { useSelector } from "react-redux";
import {
  FormControl,
  InputLabel,
  Select,
  MenuItem,
  Checkbox,
  ListItemText,
  TextField,
  Typography,
  InputAdornment,
  Button
} from "@material-ui/core";
import SaveIcon from "@material-ui/icons/Save";

const SplitPayment = ({ availableJars, amount, setSplitJars }) => {
  const { defaultJar } = useSelector(state => state);
  const [availableJarsState] = useState(
    availableJars.map(jar => ({
      ...jar,
      amount: 0,
      percent: 0
    }))
  );
  const [choosedJars, setChoosedJars] = useState([]);
  const [sumAmount, setSumAmount] = useState(amount);
  useEffect(() => {
    setSumAmount(amount);
  }, [amount]);
  const handleChange = event => {
    setChoosedJars(event.target.value);
  };
  const joinSelectedNames = selected => {
    const names = selected.map(select => select.name);
    return names.join(", ");
  };
  const setValue = (index, value) => {
    const sum = choosedJars.reduce(
      (prevValue, jar, idx) =>
        index !== idx ? prevValue + jar.percent : prevValue,
      0
    );
    const percentSum = 100 - sum;
    if (parseFloat(value) >= 0 && parseFloat(value) <= percentSum) {
      const jars = choosedJars.map((jar, idx) => {
        if (idx === index) {
          jar.percent = parseFloat(value);
          jar.amount = parseFloat((amount / 100) * parseFloat(value)).toFixed(
            2
          );
        }
        return jar;
      });
      setChoosedJars(jars);
      setSplitJars(jars);
      const sumAmount = jars.reduce(
        (prevValue, jar, idx) => parseFloat(prevValue) + parseFloat(jar.amount),
        0
      );
      const sumAmountParsed = parseFloat(
        amount - parseFloat(sumAmount)
      ).toFixed(2);
      setSumAmount(sumAmountParsed);
    }
  };
  return (
    <>
      <FormControl margin="normal" fullWidth variant="outlined">
        <InputLabel id="demo-mutiple-checkbox-label">Wybierz słoiki</InputLabel>
        <Select
          multiple
          value={choosedJars}
          onChange={handleChange}
          renderValue={selected => joinSelectedNames(selected)}
        >
          {availableJarsState.map(jar => (
            <MenuItem key={jar.id} value={jar}>
              <Checkbox checked={choosedJars.indexOf(jar) > -1} />
              <ListItemText primary={jar.name} />
            </MenuItem>
          ))}
        </Select>
      </FormControl>
      {choosedJars.length > 0 &&
        choosedJars.map((jar, index) => {
          return (
            <FormControl
              key={index}
              margin="normal"
              fullWidth
              variant="outlined"
            >
              <Typography>
                Słoik: {jar.name}
                <br />
                Kwota w słoiku: {jar.amount}
              </Typography>
              <TextField
                disabled={amount === 0}
                value={jar.percent}
                onChange={event =>
                  setValue(index, event.target.value ? event.target.value : 0)
                }
                type="number"
                label="Wartość procentowa słoika"
                variant="outlined"
                min="0"
                max="100"
                InputProps={{
                  startAdornment: (
                    <InputAdornment position="start">%</InputAdornment>
                  )
                }}
                required
              />
            </FormControl>
          );
        })}
      <FormControl margin="normal" fullWidth variant="outlined">
        {defaultJar === "" ? (
          <Typography>
            Nie posiadasz domyślnego słoika - cała kwota musi zostać
            rozdysponowana.
          </Typography>
        ) : (
          <Typography>
            Kwota nierozdysponowana trafi do domyślnego słoika.
          </Typography>
        )}

        <Button
          disabled={
            (parseFloat(sumAmount) > 0 && defaultJar === "") || amount === 0
          }
          variant="contained"
          color="primary"
          size="large"
          type="submit"
          startIcon={<SaveIcon />}
        >
          Zapisz
        </Button>
      </FormControl>
    </>
  );
};

SplitPayment.propTypes = {
  availableJars: PropTypes.array.isRequired,
  amount: PropTypes.number.isRequired,
  setSplitJars: PropTypes.func.isRequired
};

export default SplitPayment;
