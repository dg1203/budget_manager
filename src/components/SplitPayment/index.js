import SplitPayment from "./SplitPayment";

export default SplitPayment;
