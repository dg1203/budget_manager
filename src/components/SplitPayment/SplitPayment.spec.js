import React from "react";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import SplitPayment from "./SplitPayment";
import mockedStore from "../../__mocks__/mockedStore";
import store from "../../store";

const defaultProps = {
  availableJars: [
    {
      id: "asdasda",
      name: "Jar 1",
      description: "desc",
      currency: "",
      resources: []
    },
    {
      id: "asdadsadsda",
      name: "Jar 2",
      description: "desc",
      currency: "PLN",
      resources: []
    }
  ],
  amount: 200,
  setSplitJars: jest.fn()
};

const renderComponent = (store = mockedStore) =>
  render(
    <Provider store={store}>
      <SplitPayment {...defaultProps} />
    </Provider>
  );

describe("SplitPayment component", () => {
  it("renders the component", () => {
    const container = renderComponent();
    expect(container).toMatchSnapshot();
  });
  it("should render correct inforamtion about default jar if defaultJAr state", () => {
    const { getByTestId, queryByTestId } = renderComponent();
    expect(queryByTestId("info-notdefaultjar")).toBeFalsy();
    expect(getByTestId("info-defaultjar")).toBeTruthy();
  });
  it("should render correct inforamtion about default jar if not defaultJar state", () => {
    const { getByTestId, queryByTestId } = renderComponent(store);
    expect(getByTestId("info-notdefaultjar")).toBeTruthy();
    expect(queryByTestId("info-defaultjar")).toBeFalsy();
  });
});
