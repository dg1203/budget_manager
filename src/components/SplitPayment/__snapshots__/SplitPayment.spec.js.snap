// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SplitPayment component renders the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
          id="demo-mutiple-checkbox-label"
        >
          Wybierz słoiki
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        >
          <div
            aria-haspopup="listbox"
            class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
            role="button"
            tabindex="0"
          >
            <span>
              ​
            </span>
          </div>
          <input
            type="hidden"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
            style="padding-left: 8px;"
          >
            <legend
              class="PrivateNotchedOutline-legend-2"
              style="width: 0.01px;"
            >
              <span>
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
      >
        <p
          class="MuiTypography-root MuiTypography-body1"
          data-testid="info-defaultjar"
        >
          Kwota nierozdysponowana trafi do domyślnego słoika.
        </p>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge"
          data-testid="save-button"
          tabindex="0"
          type="submit"
        >
          <span
            class="MuiButton-label"
          >
            <span
              class="MuiButton-startIcon MuiButton-iconSizeLarge"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M17 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V7l-4-4zm-5 16c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3zm3-10H5V5h10v4z"
                />
              </svg>
            </span>
            Zapisz
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
        data-shrink="false"
        id="demo-mutiple-checkbox-label"
      >
        Wybierz słoiki
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
      >
        <div
          aria-haspopup="listbox"
          class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
          role="button"
          tabindex="0"
        >
          <span>
            ​
          </span>
        </div>
        <input
          type="hidden"
          value=""
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
          style="padding-left: 8px;"
        >
          <legend
            class="PrivateNotchedOutline-legend-2"
            style="width: 0.01px;"
          >
            <span>
              ​
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
    >
      <p
        class="MuiTypography-root MuiTypography-body1"
        data-testid="info-defaultjar"
      >
        Kwota nierozdysponowana trafi do domyślnego słoika.
      </p>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge"
        data-testid="save-button"
        tabindex="0"
        type="submit"
      >
        <span
          class="MuiButton-label"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeLarge"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M17 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V7l-4-4zm-5 16c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3zm3-10H5V5h10v4z"
              />
            </svg>
          </span>
          Zapisz
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
